src/parser/jobs/nin/modules/Ninki.tsx

```diff
@@ -1,178 +1,108 @@
+	}))
+
+	private ninkiModifiers = new Map<number, GaugeModifier>([
+		// Builders
+		[this.data.actions.SPINNING_EDGE.id, {action: GCD_GAIN}],
+		[this.data.actions.GUST_SLASH.id, {combo: GCD_GAIN}],
+		[this.data.actions.AEOLIAN_EDGE.id, {combo: FINISHER_GAIN}],
+		[this.data.actions.ARMOR_CRUSH.id, {combo: FINISHER_GAIN}],
+		[this.data.actions.SHADOW_FANG.id, {action: FINISHER_GAIN}],
+		[this.data.actions.DEATH_BLOSSOM.id, {action: GCD_GAIN}],
+		[this.data.actions.HAKKE_MUJINSATSU.id, {combo: GCD_GAIN}],
+		[this.data.actions.THROWING_DAGGER.id, {action: GCD_GAIN}],
+		[this.data.actions.MUG.id, {action: MUG_GAIN}],
+		[this.data.actions.MEISUI.id, {action: MEISUI_GAIN}],
+		// Spenders
+		[this.data.actions.BHAVACAKRA.id, {action: SPENDER_COST}],
+		[this.data.actions.HELLFROG_MEDIUM.id, {action: SPENDER_COST}],
+		[this.data.actions.BUNSHIN.id, {action: SPENDER_COST}],
+	])
 
-	private ninki: number = 0
-	private ninkiHistory: DataPoint[] = []
-	private wasteBySource: {[key: number]: number} = {
-		[this.data.actions.MUG.id]: 0,
-		[this.data.actions.MEISUI.id]: 0,
-		[this.data.actions.SPINNING_EDGE.id]: 0,
-		[this.data.actions.GUST_SLASH.id]: 0,
-		[this.data.actions.AEOLIAN_EDGE.id]: 0,
-		[this.data.actions.SHADOW_FANG.id]: 0,
-		[this.data.actions.ARMOR_CRUSH.id]: 0,
-		[this.data.actions.DEATH_BLOSSOM.id]: 0,
-		[this.data.actions.HAKKE_MUJINSATSU.id]: 0,
-		[this.data.actions.THROWING_DAGGER.id]: 0,
-	}
 	private erroneousFrogs: number = 0 // This is my new band name
 
-	protected override init() {
-		this.addEventHook('cast', {by: 'player', abilityId: this.ninkiGcds}, this.onBuilderCast)
-		this.addEventHook('combo', {by: 'player', abilityId: NINKI_COMBOS.map(key => this.data.actions[key].id)}, this.onBuilderCast)
-		this.addEventHook('cast', {by: 'player', abilityId: NINKI_OGCDS.map(key => this.data.actions[key].id)}, this.onBuilderCast)
-		this.addEventHook('cast', {by: 'pet'}, this.onBunshinHit)
-		this.addEventHook('cast', {by: 'player', abilityId: NINKI_SPENDERS.map(key => this.data.actions[key].id)}, this.onSpenderCast)
-		this.addEventHook('normaliseddamage', {by: 'player', abilityId: this.data.actions.HELLFROG_MEDIUM.id}, this.onHellfrogAoe)
-		this.addEventHook('death', {to: 'player'}, this.onDeath)
-		this.addEventHook('complete', this.onComplete)
-	}
+	override initialise() {
+		super.initialise()
 
-	private onBuilderCast(event: CastEvent | LegacyComboEvent) {
-		const abilityId = event.ability.guid
-		// The .get() should never be undefined but we must appease the ts lint gods
-		this.addNinki(this.actionNinkiGain.get(abilityId) ?? 0, abilityId)
+		const playerFilter = filter<Event>().source(this.parser.actor.id)
+		const pets = this.parser.pull.actors.filter(actor => actor.owner === this.parser.actor).map(actor => actor.id)
+		this.addEventHook(playerFilter.type(oneOf(['action', 'combo'])).action(oneOf(Array.from(this.ninkiModifiers.keys()))), this.onGaugeModifier)
```

Comment:
I kind of mentioned this in passing in one of the dev channels in a conversation with acc, but this really isn't a great pattern, since you wind up hooking too many events for builders that can only build on action.  Especially once you start accounting for builders that require a successful damage event, it's probably better to split these hooks to make sure the  hooks only get the needed event type for each builder.

Reply:
I'll split them out, since I'll be changing `action` to `damage` anyway. At least for the damaging builders, the spenders and Meisui can stay as `action`.